### DISCUSSION TITLE
Advent of code in Toit!

### DISCUSSION THREAD

**Discussion by lask** [toitlang/toit#53] [Show and tell]
I want to do this year's advent of code puzzles in Toit. I have created this repository for my solutions: https://github.com/lask/toit-aoc-2021

So far I have just been running my solutions on the VM on Linux, but I want to start running them on an ESP32.

If you find a toiter way for one of my solutions, do let me know!

**Comment by erikcorry** [comment]
OK, here are my submissions.

https://github.com/erikcorry/advent-toit-2021

Spoiler alert - if you want to know how to read files in Toit, but you don't want the actual answer spoiled, just look at the template.toit file, don't look at the numbered toit files.

**Reply by erikcorry** [reply]
I won't upload my solution for day 3 yet, but I can share an issue I ran into.

'"foo".split "f"' will split into '""' and '"oo"'.  This is the same in other languages where a match at position zero results in an empty string as the first split result.

The surprising thing in Toit is that this is consistent, even if the splitting string is empty.  This means that `"foo".split ""` results in `["", "f", "o", "o"]`.  I guess that consistency shouldn't surprise me, especially since I was the one that  wrote this feature originally, but it did.

In the end it was easier to just use indexing into the string with `foo[i]` which returns integers `'0'` and `'1'` (Unicode code points 48 and 49).

**Reply by lask** [reply]
That sounds great, but if it is consistent, then why doesn't `"foo".split ""` return `["", "f", "o", "o", ""]`? After all `"baz".split "z"` returns `["ba", ""]`.

**Reply by erikcorry** [reply]
Just going to change split so that the empty separator divides a string into valid characters with no empty strings at either end.  Perhaps inconsistent, but so was the old behaviour, and at least now it's useful.

**Comment by lask** [comment]
My solution to day 3 was... horrible. 100 loc for both parts.

**Reply by erikcorry** [reply]
I'm not trying to minimize the number of lines.  I want it to be reasonably readable.

**Reply by erikcorry** [reply]
But if I was, I would be proud of the obscurity of https://github.com/erikcorry/advent-toit-2021/blob/main/day3-2-tricky.toit

**Comment by erikcorry** [comment]
Day 4 I put both parts in one program.  Already pushed to github because I don't think anyone will have problems avoiding spoilers.

**Comment by lask** [comment]
Added day 7: https://github.com/lask/toit-aoc-2021/tree/master/7
`repeat` and `reduce` made it a pretty simple task.

**Comment by erikcorry** [comment]
Working on a graphical version of the Day 1 challenge:
https://twitter.com/erikcorry/status/1469463849409695754
